sql/schemachanger: identify which clause in a dependency rule didn’t match

spilchen opened the pull request:
Before this change, debugging why a dependency rule for the Declarative Schema Changer didn’t apply was challenging. This update simplifies debugging by identifying the specific clause in each rule that didn’t qualify.

Here’s how the debug process works. Suppose we mistakenly swapped the `to` and `from` variables while checking the `TargetStatus` in this dependency rule:

```
       registerDepRule(
                "column type is changed to public after doing validation of a transient check constraint",
                scgraph.SameStagePrecedence,
                "transient-check-constraint", "column-type",
                func(from, to NodeVars) rel.Clauses {
                        colID := rel.Var("columnID")
                        return rel.Clauses{
                                from.Type((*scpb.CheckConstraint)(nil)),
                                to.Type((*scpb.ColumnType)(nil)),
                                JoinOnDescID(from, to, "table-id"),
                                to.El.AttrEqVar(screl.ColumnID, colID),
                                from.ReferencedColumnIDsContains(colID),
-                               from.TargetStatus(scpb.Transient),
-                               to.TargetStatus(scpb.ToPublic),
+                               to.TargetStatus(scpb.Transient),
+                               from.TargetStatus(scpb.ToPublic),
                                from.CurrentStatus(scpb.Status_TRANSIENT_VALIDATED),
                                to.CurrentStatus(scpb.Status_PUBLIC),
                        }
                },
        )
 }
```

During a schema change job, we might find that this rule no longer applies. To identify which clause in the rule failed, enable tracing:

```
SET tracing = on;
```

Run your schema change as usual. I ran it with `EXPLAIN` to avoid executing the job and instead focus on building the dependency graph using the rules:

```
EXPLAIN (ddl) ALTER TABLE t1 ALTER COLUMN c1 TYPE smallint;
```

After completion, check the `cockroach.log` for trace information. Search for the rule by name, and you’ll see entries like these:

```
I240925 15:52:36.394784 5910 sql/schemachanger/scplan/internal/rules/registry.go:59 ⋮ [T2,Vdemoapp,n1,client=127.0.0.1:53428,hostssl,user=‹demo›] 1131 applying dep rule "column type is changed to public after doing validation of a transient check constraint", 0 results found, took 25.938µs
I240925 15:52:36.394863 5910 sql/schemachanger/scplan/internal/rules/registry.go:73 ⋮ [T2,Vdemoapp,n1,client=127.0.0.1:53428,hostssl,user=‹demo›] 1132 dep rule "column type is changed to public after doing validation of a transient check constraint" did not apply. The first unsatisfied clause is: ‹$column-type-Target[TargetStatus] = TRANSIENT_ABSENT›
```

The first log shows no results were found by the rule, while the second log identifies the specific clause that didn’t work. In this case, the clause we changed is what broke the rule.

Since the logic to identify failed clauses can impact performance, this behavior only occurs when tracing is enabled.

Epic: None
Informs #127014
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/131376)
<!-- Reviewable:end -->

fqazi reviewed (approved):
@spilchen This is amazing, great work! Just every minor nits

<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

> Reviewed 13 of 13 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131376#-:-O7iUZFgCnFWChIdo2Sx:b-rvd271)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`pkg/sql/schemachanger/rel/query.go` line 111 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/131376#-O7iT2DmBjQaTgkt-gda:-O7iT2Dn2bTXivJyB0k2:b3yuuqg) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ca7c2b322443827fa9843d4bc8eca6021f8c1a2/pkg/sql/schemachanger/rel/query.go#L111)):*
> ```Go
> 	}
> 
> 	// If we are not collecting stats, then we need gather some diagnostics
> ```

Nit: Should this say are collecting stats?
___
*[`pkg/sql/schemachanger/rel/query.go` line 138 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/131376#-O7iTVV-94mKvsDpQ6At:-O7iTVV-94mKvsDpQ6Au:bto2187) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ca7c2b322443827fa9843d4bc8eca6021f8c1a2/pkg/sql/schemachanger/rel/query.go#L138)):*
> ```Go
> 	}
> 	if ec := getCachedEvalContext(); ec != nil {
> 		ec.slotResetCount = make([]int8, len(ec.slots))
> ```

I don't know how hot this path can get but would it just make sense to clear this out?


<!-- Sent from Reviewable.io -->

spilchen committed:
sql/schemachanger: identify which clause in a dependency rule didn’t match

Before this change, debugging why a dependency rule for the Declarative Schema
Changer (DSC) didn’t apply was challenging. This update simplifies debugging by
identifying the specific clause in each rule that didn’t qualify.

Here’s how the debug process works. Suppose we mistakenly swapped the `to` and
`from` variables while checking the `TargetStatus` in this dependency rule:

```
       registerDepRule(
                "column type is changed to public after doing validation of a transient check constraint",
                scgraph.SameStagePrecedence,
                "transient-check-constraint", "column-type",
                func(from, to NodeVars) rel.Clauses {
                        colID := rel.Var("columnID")
                        return rel.Clauses{
                                from.Type((*scpb.CheckConstraint)(nil)),
                                to.Type((*scpb.ColumnType)(nil)),
                                JoinOnDescID(from, to, "table-id"),
                                to.El.AttrEqVar(screl.ColumnID, colID),
                                from.ReferencedColumnIDsContains(colID),
-                               from.TargetStatus(scpb.Transient),
-                               to.TargetStatus(scpb.ToPublic),
+                               to.TargetStatus(scpb.Transient),
+                               from.TargetStatus(scpb.ToPublic),
                                from.CurrentStatus(scpb.Status_TRANSIENT_VALIDATED),
                                to.CurrentStatus(scpb.Status_PUBLIC),
                        }
                },
        )
 }
```

During a schema change job, we might find that this rule no longer applies. To
identify which clause in the rule failed, enable tracing:

```
SET tracing = on;
```

Run your schema change as usual. I ran it with `EXPLAIN` to avoid executing the
job and instead focus on building the dependency graph using the rules:

```
EXPLAIN (ddl) ALTER TABLE t1 ALTER COLUMN c1 TYPE smallint;
```

After completion, check the `cockroach.log` for trace information. Search for
the rule by name, and you’ll see entries like these:

```
I240925 15:52:36.394784 5910 sql/schemachanger/scplan/internal/rules/registry.go:59 ⋮ [T2,Vdemoapp,n1,client=127.0.0.1:53428,hostssl,user=‹demo›] 1131 applying dep rule "column type is changed to public after doing validation of a transient check constraint", 0 results found, took 25.938µs
I240925 15:52:36.394863 5910 sql/schemachanger/scplan/internal/rules/registry.go:73 ⋮ [T2,Vdemoapp,n1,client=127.0.0.1:53428,hostssl,user=‹demo›] 1132 dep rule "column type is changed to public after doing validation of a transient check constraint" did not apply. The first unsatisfied clause is: ‹$column-type-Target[TargetStatus] = TRANSIENT_ABSENT›
```

The first log shows no results were found by the rule, while the second log
identifies the specific clause that didn’t work. In this case, the clause we
changed is what broke the rule.

Since the logic to identify failed clauses can impact performance, this
behavior only occurs when tracing is enabled.

Epic: None
Informs #127014
Release note: None

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131376)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @fqazi)

___
*[`pkg/sql/schemachanger/rel/query.go` line 111 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/131376#-O7iT2DmBjQaTgkt-gda:-O7idneU1otIECAupbGa:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ca7c2b322443827fa9843d4bc8eca6021f8c1a2/pkg/sql/schemachanger/rel/query.go#L111)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Nit: Should this say are collecting stats?
</blockquote></details>

Done
___
*[`pkg/sql/schemachanger/rel/query.go` line 138 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/131376#-O7iTVV-94mKvsDpQ6At:-O7idpUFD-EmVzERsGnu:b-b8hami) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ca7c2b322443827fa9843d4bc8eca6021f8c1a2/pkg/sql/schemachanger/rel/query.go#L138)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

I don't know how hot this path can get but would it just make sense to clear this out?
</blockquote></details>

Good point. Resetting each entry to zero would be more efficient. I'll make that change.


<!-- Sent from Reviewable.io -->

fqazi reviewed (approved):
> Reviewed 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131376)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @spilchen)



<!-- Sent from Reviewable.io -->

spilchen commented:
TFTR! @fqazi 

bors r=fqazi

Dedej-Bergin reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131376#-:-O7j3qUw-Oo7phdnbfSl:bnfp4nl)* status: <span>:shipit:</span> complete! 2 of 0 LGTMs obtained (waiting on @spilchen)



<!-- Sent from Reviewable.io -->
